Feature skip data iteration when caching scoring

BenjaminBossan opened the pull request:
This is the proposed fix for #552.

As discussed there, my proposal is to cache `net.forward_iter`, so that we can skip the iteration over the data.

It would be very helpful if someone could think of any unintended side-effects this change could have. E.g., in theory, some code could rely on iterating over the data even in case of caching, but I can hardly imagine this happening in reality.

Also, I cannot test on GPU for the moment. I don't see how this would affect the outcome, but it could still be nice of someone verifies it works.

I deprecated `cache_net_infer` because it was not "private". However, I made the new `_cache_net_forward_iter` private.

marrrcin commented:
Can I tests this from some nightly build or do I have to build the package on my own?

BenjaminBossan commented:
@marrrcin You would need to install from source, but it's not difficult: https://github.com/skorch-dev/skorch#from-source

ottonemo commented on skorch/net.py:
```diff
@@ -911,10 +911,12 @@ def forward_iter(self, X, training=False, device='cpu'):
+    def _forward_output(self, yp, device):
+        if isinstance(yp, tuple):
+            return tuple(n.to(device) for n in yp)
+        return yp.to(device)
```
This is structurally very similar to `skorch.utils.to_tensor`. Maybe we should introduce a `skorch.utils.to_device` instead? This might also become handy if we support multiple GPUs in the future.

BenjaminBossan replied:
Good point, I moved this to skorch.utils.to_device

BenjaminBossan commented:
@ottonemo I addressed your comment, pls review again.

Before merging this, I should we consider making a new release? I wouldn't mind this new feature being only on master for some time in case it creates some trouble down the line.
